#Linear regression using pytorch built-in 

let's begain with importing the torch.nn package from pytorch , which contains the utility classes for building neural network

In [ ]:
import torch.nn as nn

In [ ]:
import torch

as before we represented the inputs and targets and matrixs 

In [ ]:
import numpy as np


In [ ]:
inputs = np.array([[73, 67, 43], 
                   [91, 88, 64], 
                   [87, 134, 58], 
                   [102, 43, 37], 
                   [69, 96, 70], 
                   [74, 66, 43], 
                   [91, 87, 65], 
                   [88, 134, 59], 
                   [101, 44, 37], 
                   [68, 96, 71], 
                   [73, 66, 44], 
                   [92, 87, 64], 
                   [87, 135, 57], 
                   [103, 43, 36], 
                   [68, 97, 70]], 
                  dtype='float32')

# Targets (apples, oranges)
targets = np.array([[56, 70], 
                    [81, 101], 
                    [119, 133], 
                    [22, 37], 
                    [103, 119],
                    [57, 69], 
                    [80, 102], 
                    [118, 132], 
                    [21, 38], 
                    [104, 118], 
                    [57, 69], 
                    [82, 100], 
                    [118, 134], 
                    [20, 38], 
                    [102, 120]], 
                   dtype='float32')

inputs = torch.from_numpy(inputs)
targets = torch.from_numpy(targets)

In [ ]:
inputs

tensor([[ 73.,  67.,  43.],
        [ 91.,  88.,  64.],
        [ 87., 134.,  58.],
        [102.,  43.,  37.],
        [ 69.,  96.,  70.],
        [ 74.,  66.,  43.],
        [ 91.,  87.,  65.],
        [ 88., 134.,  59.],
        [101.,  44.,  37.],
        [ 68.,  96.,  71.],
        [ 73.,  66.,  44.],
        [ 92.,  87.,  64.],
        [ 87., 135.,  57.],
        [103.,  43.,  36.],
        [ 68.,  97.,  70.]])

We are using 15 training examples to illustrate how to work with large datasets in small batches. 

#Dataset and DataLoader 
we'll create a tensor dataset , which allows access to rows of inputs and target as tuples and provides standard API's for working with many different types of datasets 

In [ ]:
from torch.utils.data import TensorDataset

In [ ]:
#define dataset 
train_ds=TensorDataset(inputs,targets)
train_ds[:3]

(tensor([[ 73.,  67.,  43.],
         [ 91.,  88.,  64.],
         [ 87., 134.,  58.]]), tensor([[ 56.,  70.],
         [ 81., 101.],
         [119., 133.]]))

the TensorDataset allows us to access small section of the  training data using the array indexing notation ([:3] in the above code). it returns the tuples with two elements . the 1st element contains the input variables for the selected row and the second contains the target 

we'll also create a dataLoader , which can split the data into batches of a predefined size while training . It also provides other utilites like surffing and random sampling of the data 

In [ ]:
from torch.utils.data import DataLoader

In [ ]:
#define dataLoader 
batch_size=5
train_dl=DataLoader(train_ds,batch_size,shuffle=True )

we can use the dataLoader in a for loop 

In [ ]:
inputs

tensor([[ 73.,  67.,  43.],
        [ 91.,  88.,  64.],
        [ 87., 134.,  58.],
        [102.,  43.,  37.],
        [ 69.,  96.,  70.],
        [ 74.,  66.,  43.],
        [ 91.,  87.,  65.],
        [ 88., 134.,  59.],
        [101.,  44.,  37.],
        [ 68.,  96.,  71.],
        [ 73.,  66.,  44.],
        [ 92.,  87.,  64.],
        [ 87., 135.,  57.],
        [103.,  43.,  36.],
        [ 68.,  97.,  70.]])

In [ ]:
for xb,yb in train_dl:
  print(xb)
  print(yb)
  break

tensor([[ 87., 135.,  57.],
        [102.,  43.,  37.],
        [ 92.,  87.,  64.],
        [ 91.,  87.,  65.],
        [ 91.,  88.,  64.]])
tensor([[118., 134.],
        [ 22.,  37.],
        [ 82., 100.],
        [ 80., 102.],
        [ 81., 101.]])


by comparing the inputs with the batches (xb,yb)create above we can see that the data is shuffle (shuffle=True)

in each itrration the dataloader returns one batch of data with the given batch size .If shuffle is set to True ,it shuffle the data before creating batches .Shuffling helps randamize the input to the optimization algorithum .Leading to the fast reduction in the loss 

the more there is randamization , the faster the model trains 

##nn.Linear 
instead of initializing the weights & baises manually , we can define the model using nn.Linear class from pyTorch , which does it automatically 

In [ ]:
#define model 
model=nn.Linear(3,2)#3=inputs(temp,rainfall,humidity)2=output(yeild of apples ,yeild of oranges )
print(model.weight)
print(model.bias)

Parameter containing:
tensor([[-0.4978, -0.1771,  0.1230],
        [ 0.3177, -0.2521,  0.5302]], requires_grad=True)
Parameter containing:
tensor([-0.1069,  0.5339], requires_grad=True)


pytorch model also have a helpfull .paremeters method , which returns a list containing all the weights and baises matrix present in the model .for our linear regression model , we have one weight matrix and one baises matrix  

In [ ]:
#parameters
list(model.parameters())

[Parameter containing:
 tensor([[-0.4978, -0.1771,  0.1230],
         [ 0.3177, -0.2521,  0.5302]], requires_grad=True),
 Parameter containing:
 tensor([-0.1069,  0.5339], requires_grad=True)]

we can use the model to genarate pridiction in same way as before 

In [ ]:
#genarate pridiction 
preds=model(inputs )
preds

tensor([[-43.0214,  29.6351],
        [-53.1174,  41.1938],
        [-60.0084,  25.1451],
        [-53.9466,  41.7188],
        [-42.8441,  35.3679],
        [-43.3422,  30.2050],
        [-52.8174,  41.9761],
        [-60.3833,  25.9930],
        [-53.6259,  41.1490],
        [-42.2233,  35.5803],
        [-42.7214,  30.4174],
        [-53.4382,  41.7637],
        [-60.3085,  24.3629],
        [-54.5674,  41.5064],
        [-42.5233,  34.7981]], grad_fn=<AddmmBackward>)

##Loss Function 
instead of defining loss function  manually we can use the built-in loss function "mse_loss"

In [ ]:
# importing nn function 
import torch.nn.functional as F

the nn.functional  package contains many usefull loss function and several other utilities   

In [ ]:
#define loss function 
loss_fn=F.mse_loss

let's compute the loss for the current prediction of our  model 

In [ ]:
loss=loss_fn(model(inputs),targets)
print(loss)

tensor(11035.3281, grad_fn=<MseLossBackward>)


##optimizer
instead of manuallly manupalating the models's weights and baises using gradiants, we can use the optimizer optim.SGD (stochastic gradiant decent ).Stochastic indicates that samples are selected in random batches instead of a single group 

In [ ]:
#defining optimizer
opt=torch.optim.SGD(model.parameters(),lr=1e-5)

Note that the model.parameters() is passed as an argument to optim.SGD so that the optimizer knows which matrix to be modified during the update step .also we can specify a learning rate that contraols the amount by which the parameters are modified  

## Train the model
we are now ready to train the model . We'll follow the same steps to implement gradiant decent 

*   genarate prediction
*   calculate the loss 
*   compute gradient w.r.t weights and baises 
*   adjust the weights by substracting a small quantity proportional to gradient 
*   rest the gradient to zero 

the only change is that we'll work on batches of data instead of processing the entire  training data  in every itration . Let's define a utility function fit that train the model for a given number of epochs 






In [ ]:
#utility function to train the model 
def fit (num_epochs, model,loss_fn,opt, train_dl):

  #repeat for the given number of epochs 
  for epoch in range (num_epochs):

    #train the batchs of data 
    for xb, yb in train_dl:

      #1 genarate pridiction 
      pred=model(xb)

      #2 calculate the loss 
      loss=loss_fn(pred,yb)

      #3 compute gradiant 
      loss.backward()

      #4 update parameters using gradiant 
      opt.step()

      #5 reset the gradiant to zero
      opt.zero_grad()

      #6 print the progress
      if(epoch+1)  % 10==0:
        print("Epoch [{},{}] , Loss :{:.4f", format(epoch+1), num_epochs,loss.item())




### somethings to note above
*   we use the data loader defined earlier to get batches of data for every itration.
*   instead of updating parameters(weights & baises) manualy we use opt.step to perform the update and opt.zero_grad to reset the gradiant to zero
*    we also add the log statment that prints the loss from the last batch of data for every 10th epoch to track the traning progress . loss.item retures the actual value stored in the loss tensor 

Let's train the model for 100 epochs



In [ ]:
fit(100,model,loss_fn,opt,train_dl)

Epoch [{},{}] , Loss :{:.4f 10 100 262.38519287109375
Epoch [{},{}] , Loss :{:.4f 10 100 463.1507263183594
Epoch [{},{}] , Loss :{:.4f 10 100 105.70985412597656
Epoch [{},{}] , Loss :{:.4f 20 100 67.32218933105469
Epoch [{},{}] , Loss :{:.4f 20 100 202.79148864746094
Epoch [{},{}] , Loss :{:.4f 20 100 301.6537780761719
Epoch [{},{}] , Loss :{:.4f 30 100 118.35737609863281
Epoch [{},{}] , Loss :{:.4f 30 100 208.3970184326172
Epoch [{},{}] , Loss :{:.4f 30 100 67.3210678100586
Epoch [{},{}] , Loss :{:.4f 40 100 95.81165313720703
Epoch [{},{}] , Loss :{:.4f 40 100 70.26551818847656
Epoch [{},{}] , Loss :{:.4f 40 100 96.84049987792969
Epoch [{},{}] , Loss :{:.4f 50 100 69.88969421386719
Epoch [{},{}] , Loss :{:.4f 50 100 53.0590705871582
Epoch [{},{}] , Loss :{:.4f 50 100 57.57508087158203
Epoch [{},{}] , Loss :{:.4f 60 100 38.878204345703125
Epoch [{},{}] , Loss :{:.4f 60 100 20.244686126708984
Epoch [{},{}] , Loss :{:.4f 60 100 67.06303405761719
Epoch [{},{}] , Loss :{:.4f 70 100 42.5747

In [ ]:
preds=model(inputs)
preds

tensor([[ 57.6856,  71.7047],
        [ 81.9604, 101.0835],
        [117.6580, 129.9747],
        [ 24.7484,  44.3994],
        [ 99.3864, 115.6383],
        [ 56.5330,  70.8100],
        [ 81.7414, 101.1898],
        [117.9404, 130.6281],
        [ 25.9010,  45.2941],
        [100.3199, 116.6393],
        [ 57.4665,  71.8110],
        [ 80.8079, 100.1888],
        [117.8770, 129.8684],
        [ 23.8149,  43.3983],
        [100.5389, 116.5330]], grad_fn=<AddmmBackward>)

In [ ]:
targets

tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.],
        [ 57.,  69.],
        [ 80., 102.],
        [118., 132.],
        [ 21.,  38.],
        [104., 118.],
        [ 57.,  69.],
        [ 82., 100.],
        [118., 134.],
        [ 20.,  38.],
        [102., 120.]])

indeed ,the prediction are close to the target. we have a trained a resonably good model to pridict crop yeild for apples and oranges by looking at the avgrage temperature , rainfall,and humidity in a reign. we can use this to make pridiction of crop yeild for new regions by passinf a batch of single row of inputs 

In [ ]:
model(torch.tensor([[75,63,44.]]))

tensor([[54.3343, 69.3008]], grad_fn=<AddmmBackward>)

the pridicted yeild of apple is 54.33 tons per hector and that of oranges is 69.30 tons per hector 

## Machine Learning vs. Classical Programming

 we've defined a "model" that assumes a specific relationship between the inputs and the outputs, expressed using some unknown parameters (weights & biases). We then show the model some know inputs and outputs and _train_ the model to come up with good values for the unknown parameters. Once trained, the model can be used to compute the outputs for new inputs.

This paradigm of programming is known as _machine learning_, where we use data to figure out the relationship between inputs and outputs. _Deep learning_ is a branch of machine learning that uses matrix operations, non-linear activation functions and gradient descent to build and train models. Andrej Karpathy, the director of AI at Tesla Motors, has written a great blog post on this topics, titled [Software 2.0](https://medium.com/@karpathy/software-2-0-a64152b37c35).

This picture from book [Deep Learning with Python](https://www.manning.com/books/deep-learning-with-python) by Francois Chollet captures the difference between classical programming and machine learning:

![](https://i.imgur.com/oJEQe7k.png)

Keep this picture in mind as you work through the next few tutorials. 
